src/services/codefixes/convertFunctionToEs6Class.ts

```diff
@@ -102,51 +141,76 @@ namespace ts.codefix {
-                        const method = createMethod(/*decorators*/ undefined, fullModifiers, /*asteriskToken*/ undefined, memberDeclaration.name, /*questionToken*/ undefined,
-                            /*typeParameters*/ undefined, functionExpression.parameters, /*type*/ undefined, functionExpression.body);
-                        copyLeadingComments(assignmentBinaryExpression, method, sourceFile);
-                        return method;
-                    }
-
-                    case SyntaxKind.ArrowFunction: {
-                        const arrowFunction = assignmentBinaryExpression.right as ArrowFunction;
-                        const arrowFunctionBody = arrowFunction.body;
-                        let bodyBlock: Block;
-
-                        // case 1: () => { return [1,2,3] }
-                        if (arrowFunctionBody.kind === SyntaxKind.Block) {
-                            bodyBlock = arrowFunctionBody as Block;
-                        }
-                        // case 2: () => [1,2,3]
-                        else {
-                            bodyBlock = createBlock([createReturn(arrowFunctionBody)]);
-                        }
-                        const fullModifiers = concatenate(modifiers, getModifierKindFromSource(arrowFunction, SyntaxKind.AsyncKeyword));
-                        const method = createMethod(/*decorators*/ undefined, fullModifiers, /*asteriskToken*/ undefined, memberDeclaration.name, /*questionToken*/ undefined,
-                            /*typeParameters*/ undefined, arrowFunction.parameters, /*type*/ undefined, bodyBlock);
-                        copyLeadingComments(assignmentBinaryExpression, method, sourceFile);
-                        return method;
+                    case SyntaxKind.FunctionExpression:
+                    case SyntaxKind.ArrowFunction:
+                        return createFunctionLikeExpressionMember(members, <FunctionExpression | ArrowFunction>assignmentBinaryExpression.right, (<PropertyAccessExpression>memberDeclaration).name);
+                    case SyntaxKind.ObjectLiteralExpression: {
+                        return map(
+                            (<ObjectLiteralExpression>assignmentBinaryExpression.right).properties,
+                            property => {
+                                if (isMethodDeclaration(property) || isGetOrSetAccessorDeclaration(property)) {
+                                    // MethodDeclaration and AccessorDeclaration can appear in a class directly
+                                    return members.concat(property);
+                                }
+                                // Drop constructor assignments
+                                if (isConstructorAssignment(property)) return members;
+                                return createFunctionLikeExpressionMember(members, <FunctionExpression | ArrowFunction>(<PropertyAssignment>property).initializer, property.name!);

```

Comment:
At a glance I’m having trouble following all the type assertions through this section. The `ObjectLiteralExpression` one and the first `FunctionExpression | ArrowFunction` one are obvious enough, but it’s hard to see why the `PropertyAccessExpression` one and `<FunctionExpression | ArrowFunction>(<PropertyAssignment>property).initializer` are safe. I’m guessing it’s due to the checks in `shouldConvertDeclaration`. I would recommend creating a type that represents every kind of declaration that can be converted, and then turning `shouldConvertDeclaration` into a type predicate returning `expression is ...` or `_target is ...`. Then, this `switch` statement on `assignmentBinaryExpression.right.kind` could become a series of `if` statements using other existing type guards like `isFunctionLikeExpression` and `isObjectLiteralExpression`... that would make the types narrow through the control flow and would eliminate the need for most of these type assertions.

Reply:
Hmm, that's another piece of the original code, but fixed.